clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -1422,6 +1424,9 @@ Value *ScalarExprEmitter::EmitScalarConversion(Value *Src, QualType SrcType,
     return Builder.CreateVectorSplat(NumElements, Src, "splat");
   }
 
+  if (SrcType->isExtVectorType() && DstType->isExtVectorType())
+    return EmitVectorElementConversion(SrcType, DstType, Src);
```

Comment:
EmitScalarConversion is a mess; we have cast kinds, but we never taught this part of CodeGen to use them, so it just guesses the cast kind based on the source and destination types.  This is particularly bad for vectors, since it's often possible to cast them in multiple different ways; I'm not confident this change won't have unintended side-effects.  Please fix the code for the relevant cast kinds to avoid going through EmitScalarConversion.

Reply:
Yeah, I believe the reason we still have this whole analysis in `EmitScalarConversion` is, sadly enough, just to deal with the LHS and final result of a compound assignment expression, which we don't have `CastKind`s for.  The more we can get away from relying on this, the better.